### ISSUE TITLE
Make the Gradle distribution slimmer

### ISSUE THREAD

**Issue by melix** [gradle/gradle#3513]
The Gradle distribution zip file is way too fat. There are lots of things most people would never use in there. Right now, for each distribution, we download a 76MB zip file. In this zip, we can find:

- the Kotlin compiler and runtime, 30MB, even if 99% of our users don't use Kotlin scripts
- `bcprov-jdk15on-1.58.jar`, 4MB, which is suspicious and the fact it says jdk15 doesn't make me super confident why it's here
- bnd, only used if one applies the `osgi` plugin
- jgit, why?
- ...

In short, we pay an excessive price for the distribution, where we could be lazy. The good news is that we most likely know the coordinates of those files, and we also have a dependency management engine. So we could probably download most of them _on demand_, and reuse our artifact cache to save some precious space.

**Comment by oehme** [comment]
JGit is for external source dependencies.

I'd be happy if you could deprecate the OSGi plugin, pointing users to the excellent BND plugin instead. Then we can get rid of the dependency in 5.0. We'd need to do it now though.

**Comment by melix** [comment]
It's less about what we can remove/deprecate, than it is about _lazily downloading_ dependencies IMO.

**Comment by oehme** [comment]
We should deprecate OSGi, then we don't need any smart on-demand dependency download for it.

For some of the other things I agree that on-demand would be better. Kotlin seems like the most obvious candidate.

**Comment by smoothreggae** [comment]
`bcprov-jdk15on-1.58.jar` is used by the [signing plugin](https://docs.gradle.org/current/userguide/signing_plugin.html). The confusing suffix `jdk15on` stems from the Bouncy Castle folks naming their releases based on the targeted JVM levels. `jdk15` was the suffix for JDK 5. They didn't have any changes specific to any higher JDK levels (1.6 through 1.8) so they stayed with the suffix `jdk15`. They changed this suffix ever so slightly in newer releases from `jdk15` to `jdk15on` (I believe it means "JDK 1.5 and onward").

**Comment by jjohannes** [comment]
Make distribution smaller by moving plugins out: #9428
